_posts/2021-02-03-#20721.md

```diff
@@ -0,0 +1,111 @@
+    [net.cpp](https://github.com/bitcoin/bitcoin/blob/16b784d9/src/net.h). Net
+    hands messages up the stack to ...
+
+  - **Net Processing** - reads and processes application-level messages, and
+    handles application-level view of peers. The main class in net
+    processing is `PeerManager`, which is declared in
+    [net_processing.h](https://github.com/bitcoin/bitcoin/blob/16b784d9/src/net_processing.h).
+    Net processing uses function calls like `ProcessNewBlock()` and
+    `AcceptToMemoryPool()` to call into ...
+
+  - **Validation** - is responsible for our view of the block chain and the UTXO
+    set. It may (or may not) include a mempool for unconfirmed transactions. There
+    are various classes and functions exposed to the rest of the node. [PR
+    20158](https://github.com/bitcoin/bitcoin/pull/20158) and linked PRs are an
+    attempt to clarify and rationlize the interface to validation.
+
+- In early versions of Bitcoin, this distinction wasn't as clear. The [first
+  commit](https://github.com/jnewbery/bitcoin/blob/4405b78d/main.cpp) in the
+  Bitcoin Core repository has a separate net.cpp file, but net processing and
+  validation are mixed in the main.cpp file. It was only in [PR
+  9260](https://github.com/bitcoin/bitcoin/pull/9260) in 2016 that net processing
+  and validation were split into separate files.
+
+- Even today, the division between net, net processing and validation is not as
+  clear as we'd like. Clarifying and rationalizing these interfaces is an
+  ongoing project. See, for example, [PR
+  15141](https://github.com/bitcoin/bitcoin/pull/15141), which moves the denial
+  of service logic from validation into net processing, [PR
+  20158](https://github.com/bitcoin/bitcoin/pull/20158), which rationalizes the
+  interface to validation, and [PR
+  19398](https://github.com/bitcoin/bitcoin/issues/19398), which moves
+  application-layer data and logic from net into net processing.
+
+- The high level goal of PR 19398 is to split responsibilities cleanly between
+  CConnman and PeerManager:
+
+  - C**Conn**man should be responsible for **connections** - accepting incoming
+    and opening outbound connections, opening and closing sockets, reading and
+    writing data on those sockets, and first-line-of-defence validation of
+    corrupt/malformed messages.
+
+  - **Peer**Manager should be responsible for **peers** - parsing and acting on
+    p2p messages from peers, storing state relating to those peers, and
+    deciding when to send messages to those peers.
+
+- Today's PR, [PR 20721](https://github.com/bitcoin/bitcoin/issues/20721), is
+  part of the larger project tracked by PR 19398. It moves the `ping` data
+  from net into net processing.
+
+## Notes
+
+- The `ping`/`pong` message exchange acts as a keepalive between peers on the
+  peer-to-peer network, and also measures round-trip latency. Best ping time
+  and most recent ping time are presented to users in the GUI and RPC interface,
```

Comment:
GUI, RPC and CLI ;)

Reply:
Hmmm, the CLI is a separate client which uses the RPC interface, it's not an interface itself.